RemoteResourceCacheProxy::clearDecomposedGlyphsMap should deregister observers

heycam opened the pull request:
#### d1b1809daffb4b06b6626c72be82bf28e5bd2881
<pre>
RemoteResourceCacheProxy::clearDecomposedGlyphsMap should deregister observers
<a href="https://bugs.webkit.org/show_bug.cgi?id=242325">https://bugs.webkit.org/show_bug.cgi?id=242325</a>
rdar://96209023

Reviewed by Myles C. Maxfield.

Similar to clearNativeImageMap, clearDecomposedGlyphsMap should
deregister the DecomposedGlyphs observers when we clear out
m_decomposedGlyphs, since we no longer need to observe them.

* LayoutTests/fast/text/glyph-display-lists/glyph-display-list-gpu-process-crash-expected.txt: Added.
* LayoutTests/fast/text/glyph-display-lists/glyph-display-list-gpu-process-crash.html: Added.
* Source/WebCore/rendering/TextPainter.cpp:
(WebCore::TextPainter::clearGlyphDisplayListCacheForTesting):
* Source/WebCore/rendering/TextPainter.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::clearGlyphDisplayListCacheForTesting):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::clearDecomposedGlyphsMap):

Canonical link: <a href="https://commits.webkit.org/252170@main">https://commits.webkit.org/252170@main</a>
</pre>

litherum commented on Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
```diff
@@ -196,6 +196,8 @@ void RemoteResourceCacheProxy::clearImageBufferBackends()
 void RemoteResourceCacheProxy::clearDecomposedGlyphsMap()
 {
+    for (auto& decomposedGlyphs : m_decomposedGlyphs.values())
+        decomposedGlyphs->removeObserver(*this);
```
`m_decomposedGlyphs` holds `WeakPtr`s. Do we need to do `null` checks here?

Is `this` guaranteed to be the only observer?

The commit message doesn't have any explanation about why the current behavior of `DecomposedGlyphs::~DecomposedGlyphs()` isn't sufficient.

Can this be tested?

heycam replied:
> `m_decomposedGlyphs` holds `WeakPtr`s. Do we need to do `null` checks here?

We don't because the destructor for the `CanMakeWeakPtr` base class of `DecomposedGlyphs` runs after the `DecomposedGlyphs` destructor, and the latter will drop the `WeakPtr`s by notifying the observers.

> Is `this` guaranteed to be the only observer?

No, since different documents could pick the same entry out of the global `GlyphDisplayListCache`.

> Can this be tested?

I will think about that.

heycam committed:
RemoteResourceCacheProxy::clearDecomposedGlyphsMap should deregister observers
https://bugs.webkit.org/show_bug.cgi?id=242325
rdar://96209023

Reviewed by Myles C. Maxfield.

Similar to clearNativeImageMap, clearDecomposedGlyphsMap should
deregister the DecomposedGlyphs observers when we clear out
m_decomposedGlyphs, since we no longer need to observe them.

* LayoutTests/fast/text/glyph-display-lists/glyph-display-list-gpu-process-crash-expected.txt: Added.
* LayoutTests/fast/text/glyph-display-lists/glyph-display-list-gpu-process-crash.html: Added.
* Source/WebCore/rendering/TextPainter.cpp:
(WebCore::TextPainter::clearGlyphDisplayListCacheForTesting):
* Source/WebCore/rendering/TextPainter.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::clearGlyphDisplayListCacheForTesting):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::clearDecomposedGlyphsMap):

Canonical link: https://commits.webkit.org/252170@main
